sentry-core/src/main/java/io/sentry/core/SentryClient.java

```diff
@@ -88,43 +88,83 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
                 SentryLevel.DEBUG,
                 "Event was dropped by processor: %s",
                 processor.getClass().getName());
+        sendEvent = false;
         break;
       }
     }
 
-    if (event == null) {
-      return SentryId.EMPTY_ID;
-    }
+    Session session = null;
 
-    // TODO: should it be done on the HUB?
-    updateSessionData(event, hint, scope);
+    if (event != null) {
+      session = updateSessionData(event, hint, scope);
 
-    if (!sample()) {
-      options
-          .getLogger()
-          .log(
-              SentryLevel.DEBUG,
-              "Event %s was dropped due to sampling decision.",
-              event.getEventId());
-      return SentryId.EMPTY_ID;
+      if (!sample()) {
+        options
+            .getLogger()
+            .log(
+                SentryLevel.DEBUG,
+                "Event %s was dropped due to sampling decision.",
+                event.getEventId());
+        sendEvent = false;
+        event = null;
+      }
     }
 
-    event = executeBeforeSend(event, hint);
+    if (event != null) {
+      event = executeBeforeSend(event, hint);
 
-    if (event == null) {
-      options.getLogger().log(SentryLevel.DEBUG, "Event was dropped by beforeSend");
-      return SentryId.EMPTY_ID;
+      if (event == null) {
+        options.getLogger().log(SentryLevel.DEBUG, "Event was dropped by beforeSend");
+        sendEvent = false;
+      }
     }
 
+    SentryId sentryId = SentryId.EMPTY_ID;
+
     try {
-      connection.send(event, hint);
+      final SentryEnvelope envelope = buildEnvelope(event, session);
+
+      if (envelope != null) {
+        connection.send(envelope, hint);
+      }
```

Comment:
When would envelope be null here?
Should we log this out?

Reply:
if the event was discarded and there is no session to be sent,  the `buildEnvelope` would return null, as there's nothing to be sent

Reply:
Got it. thanks